compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -518,68 +499,124 @@ pub fn report_dyn_incompatibility<'tcx>(
         }
     }
 
+    attempt_dyn_to_enum_suggestion(tcx, trait_def_id, &*trait_str, &mut err);
+
+    err
+}
+
+/// Attempt to suggest converting the `dyn Trait` argument to an enumeration
+/// over the types that implement `Trait`.
+fn attempt_dyn_to_enum_suggestion(
+    tcx: TyCtxt<'_>,
+    trait_def_id: DefId,
+    trait_str: &str,
+    err: &mut Diag<'_>,
+) {
     let impls_of = tcx.trait_impls_of(trait_def_id);
-    let impls = if impls_of.blanket_impls().is_empty() {
-        impls_of
-            .non_blanket_impls()
-            .values()
-            .flatten()
-            .filter(|def_id| {
-                !matches!(tcx.type_of(*def_id).instantiate_identity().kind(), ty::Dynamic(..))
-            })
-            .collect::<Vec<_>>()
-    } else {
-        vec![]
+
+    // Don't suggest converting to an enum if there are any blanket impls.
+    if !impls_of.blanket_impls().is_empty() {
+        return;
+    }
+
+    let concrete_impls = {
+        let mut concrete_impls = Vec::new();
+        for impl_id in impls_of.non_blanket_impls().values().flatten() {
+            // Don't suggest converting to enum if there are any non-lifetime generics.
+            if has_non_lifetime_generics(tcx, *impl_id) {
+                return;
+            }
```

Comment:
The specific cases I was trying to avoid were suggestions on impls like `impl<T: SomeTrait> SomeTrait for &mut T { .. }`. For example, [this playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ea358a0a61ae13afdb53bc2478edc5cb) generates a note like:

```
  = help: the following types implement the trait, consider defining an enum where each variant holds one of these types, implementing `SomeTrait` for this new enum and using it instead:
            &mut T
            std::boxed::Box<T>
            u32
```

These suggestions are, in my experience, unlikely to be what is wanted.

More generally, traits with type-generic or const-generic impls are unlikely to be things that can be easily transformed into an enum, and these traits and impls are typically are written by more intermediate users who are already aware that they could instead use an enum.

I wanted to exclude lifetime parameters because it's common for lifetime parameters in impls to be conceptually "one type" (e.g. `SomeTy<'_>` is maybe something that makes sense to have as a case in an enum). I don't feel strongly about this.

I've switched to use `tcx.type_of(*impl_id).no_bound_vars()` which uses `has_param` as you suggested.